Local Variable Type Hints

APickledWalrus opened the pull request:
### Problem
Local variables are essentially accepted wherever they are used. This results in more potential mistakes and (some) strange syntax behavior that is hard to track. Detailed type information for variables helps in two cases:
#### 1. Mismatched types
It's possible you use the wrong variable somewhere. Detailed type information can catch this and print a warning (or error). Consider:
```applescript
set {_a} to 5
set {_b} to "some string"
... do stuff ...
set {_c} to {_a} in lowercase # oops i used the wrong variable
```
We can catch that `{_a} in lowercase` is likely to be invalid and alert the user.

#### 2. Syntax accepting objects
Consider syntax like ExprArithmetic:
```applescript
set {_a} to "some string"
set {_b} to 1 + {_a}
```
With detailed type information, we can now identify at parse time that this is an invalid operation, instead of failing at runtime and returning none.

### Solution
With systems like ExecutionIntent and multiple return types, we are finally in a position to execute this system well. Type tracking for local variables has been implemented across syntax interacting with variables using a HintManager that is attached to the ParserInstance.

The testing script along with HintManager's documentation provide insight into how hints are managed.

As of now, I believe this system is accurate enough for type issues to be errors rather than warnings. However, there are still scenarios where type hints will fail:
```
loop x times:
  if ...:
    set {_x} to 10
  if ...:
    # will only see {_x} as a number
    set {_x} to x in lowercase
    # do some terminating statement idk
  if ...:
    set {_x} to "hello world"
```
That is, variables in loops whose types are changing across iterations may trigger false-positive errors (in rare cases hidden type changes could cause erroneous runtime failures). To work around this, it is possible to [un]suppress type hints (for a statement or block of code) using a new statement. We may also consider putting this behind an experiment as we continue to improve upon it (maybe whether it's an error?). Variables that cannot have hints and variables that do not have any hints act the same as they do now.

Variables have also been reworked here. When variables are passed to patterns request objects, the possible return types of the variable is simply an array of the known hints. Additionally, when variables are passed to patterns requesting some combination of types, the possible return types of the variable is the intersection of the hints and requested types.

I believe the system is accurate and stable, but further testing is needed. As a result, I have elected to put this behind an experiment. Additionally, it is possible to suppress type hints (in scripts where the experiment is enabled) for specific regions of code.

**Change Explanations**
- Fixed an issue where the `experimentRegistry` (in Skript) was set to null too soon - it could be null even when more script code could run (consider `on stop` triggers)
- Added `getName` to arguments. This is used for providing type hints for named command arguments.
- Fixed an issue where EffReturn would return an incorrect ExecutionIntent for Functions. It needed to be using `StopTrigger` as the `ReturnableTrigger` is stopped in Functions when `return` is executed.
- Added `EntryNode` as a possible return type for ExprNode. This is for ensuring type hints work properly for the syntax.
- Added `beforeChange` runnable for loading sections into triggers. This was necessary for proper type hint implementation for these syntaxes. We need to be able to capture the hints and then merge them in once the trigger is loaded (since we copy variables at runtime).
- Now that some variables may have accurate return types, `Variable#getConvertedExpression` has been updated to restrict what conversions are allowed on those variables.
- Fixed an issue where `Classes#getSuperClassInfo` was not guaranteed to return the exact ClassInfo if one existed.
- I added a new class, `SectionUtils`, which includes a static utility method for loading a section's code into a Trigger. More specifically, it handles the type of behavior that is becoming more common: Loading a section into a Trigger to use different context/events, preventing delays, and copying local variables at runtime. With type hints, the need to copy type hints arose, and the behavior became far more complicated. This method serves to handle that behavior and reduce the effort needed by the increasing number of implementing classes.

#### API Support

Addon developers may need to support this feature if their syntax modifies variables (and results in type changes). First, you can use the HintManager to check if a Variable can use hints:
```java
if (HintManager.canUseHints(variable)) {
    // safe to manage hints
}
```
If so, you can obtain the active HintManager using the ParserInstance:
```java
ParserInstance.get().getHintManager();
```
Then, depending on what type of operation your syntax is performing, you can manage hints as necessary:
```java
HintManager hintManager = ParserInstance.get().getHintManager();
// setting the hints (clears previous hints)
hintManager.set(variable, SomeType.class);
// adding a hint
hintManager.add(variable, SomeType.class);
// clearing hints (e.g. `delete {_var}`)
hintManager.delete(variable);
```

### Testing Completed
I have added a basic `type hints.sk` test (more to come). I plan to implement JUnit testing for HintManager.

### Supporting Information
Depends on:
- https://github.com/SkriptLang/Skript/pull/7891

---
**Completes:**
- #3864
- #1229

**Related:**
- https://github.com/SkriptLang/Skript/pull/7879 (syntax conflicts (due to variables) would benefit from type hints)

APickledWalrus committed:
Implement basic local variable type hints

APickledWalrus committed:
Expand HintManager API

APickledWalrus committed:
Use possibleReturnTypes for better hints

APickledWalrus committed:
Test expanded Variable types

APickledWalrus committed:
Implement additional hint management

These syntaxes were giving some false positive warnings

APickledWalrus committed:
Merge branch 'dev/feature' into feature/local-variable-type-hints

APickledWalrus committed:
Clear some hints when possible

We do not want to copy hints from sections containing a statement that terminate the trigger

APickledWalrus committed:
Rework hint availability check

APickledWalrus committed:
Fix accidental modern method usage

APickledWalrus committed:
Implement more advanced scope control

APickledWalrus committed:
Implement type hint updates for more syntax

sovdeeth commented on src/main/java/ch/njol/skript/ScriptLoader.java:
```diff
@@ -951,6 +951,10 @@ public static ArrayList<TriggerItem> loadItems(SectionNode node) {
 
+		// Begin local variable type hints
+		parser.getHintManager().enterScope();
+		boolean freezeHints = false;
```
some explanation about what 'freezing hints' means would be nice

sovdeeth commented on src/main/java/ch/njol/skript/effects/EffChange.java:
```diff
@@ -234,25 +236,45 @@ else if (mode == ChangeMode.SET)
+					if (ch instanceof ExprParse exprParse) {
+						exprParse.flatten = false;
+					} else if (ch instanceof ExpressionList<?> exprList) {
+						for (Expression<?> expr : exprList.getExpressions()) {
+							if (expr instanceof ExprParse exprParse) {
+								exprParse.flatten = false;
+							}
+						}
```
note that since this is not recursive, it cannot be relied upon to flatter all the possible exprparses
see https://github.com/SkriptLang/Skript/pull/7850

APickledWalrus replied:
im fine to change it here, or someone can do a separate PR and ill just integrate it

APickledWalrus replied:
Addressed in https://github.com/SkriptLang/Skript/pull/7915

sovdeeth commented on src/main/java/ch/njol/skript/ScriptLoader.java:
```diff
@@ -1037,7 +1037,36 @@ public static ArrayList<TriggerItem> loadItems(SectionNode node) {
+				// Exiting only one section does not require special behavior
+				if (item.executionIntent() instanceof ExecutionIntent.StopSections intent && intent.levels() > 1) {
+					parser.getHintManager().mergeScope(0, intent.levels());
+					parser.getHintManager().enterScope(); // Enter a new scope to capture all new type hints
+					// Clear scope to prevent duplicate copying
+					// We clear after entering the capturing scope so that the capturing scope still has hints
+					parser.getHintManager().clearScope(1);
+				} else {
+					parser.getHintManager().enterScope(); // Enter a new scope to capture all new type hints
```
can you elaborate on why enterScope is used here instead of exitScope?
It seems to me like there are more enters than exits in an arbitrary structure

APickledWalrus replied:
The enter scope call for this condition is to capture type hints. Consider the following code:
```
# do some stuff
exit 3 sections
set {_a} to 5
```
There is type hint information about {_a}, but that code will never actually run, so we don't want to capture those type hints and pass them back up as possible types. So, once we reach a terminating statement (like EffExit), further type hints are ignored. This is accomplished by entering another scope at the beginning (which is what you see here) and at the end of `loadItems`, we clear and exit the scope (clearing it means no hints will be passed back up). This is what the freezeHints variable tracks.

Writing it out though, it might not actually be necessary. Unless the scope is `exit 1 section`, hints on the regular/current scope are already cleared and passed up early (that's what the if of this if/else handles). It's probably fine to do that for `exit 1 section` too and we can avoid this additional scope enter/exit

sovdeeth replied:
I may be misunderstanding but can't you just `mergeScope(0, target)` and freeze on exit x sections and otherwise leave it alone?

APickledWalrus committed:
Implement hints for function and command parameters

APickledWalrus committed:
Simplify ScriptLoader scope freezing logic

APickledWalrus committed:
Remove non-ideal safety checks

APickledWalrus committed:
Check that command arguments have names

APickledWalrus committed:
Copy hints for sections that copy variables

APickledWalrus committed:
Combine Variable type determination

Variables will always use converters. so this allows us to represent a broader range of types (and likely reduce mistakes)

APickledWalrus committed:
Enhance list variable hints

APickledWalrus committed:
Merge branch 'dev/feature' into feature/local-variable-type-hints

APickledWalrus committed:
Re-implement EffChange changes

APickledWalrus committed:
Type hint errors

Type hints have become more advanced, and the goal is now for them to become errors. That is, the system should be opt-out rather than opt-in. Confidence in hint accuracy is high enough.

APickledWalrus committed:
Add support for EffDoIf

APickledWalrus committed:
Merge branch 'dev/feature' into feature/local-variable-type-hints

APickledWalrus committed:
Invalid function argument test

APickledWalrus committed:
HintManager documentation tweaks

APickledWalrus committed:
Mark type hints as experimental

APickledWalrus committed:
More fixes and cleanup

ContainerExpression behavior was wrong - variables should not be container expressions

Absolutionism commented on src/main/java/ch/njol/skript/sections/EffSecShoot.java:
```diff
@@ -191,12 +195,26 @@ public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelaye
 			AtomicBoolean delayed = new AtomicBoolean(false);
-			Runnable afterLoading = () -> delayed.set(!getParser().getHasDelayBefore().isFalse());
-			trigger = loadCode(sectionNode, "shoot", afterLoading, ShootEvent.class);
+			AtomicReference<HintManager.Backup> hintBackup = new AtomicReference<>();
+			// Copy hints and ensure no delays
+			Runnable beforeLoading = () -> getParser().getHintManager().enterScope(false);
+			Runnable afterLoading = () -> {
+				delayed.set(!getParser().getHasDelayBefore().isFalse());
+				HintManager hintManager = getParser().getHintManager();
+				hintBackup.set(hintManager.backup());
+				hintManager.exitScope();
+			};
+
+			trigger = loadCode(sectionNode, "shoot", beforeLoading, afterLoading, ShootEvent.class);
+
```
Just a thought, but could there be a helper method for this? Since `EffSecShoot` , `EffSecSpawn`, `ExprSecCreateWorldBorder` and `ExprSecCreateLootContext` all have this layout?

Absolutionism replied:
Alrighty, it would be nice to have, especially with incoming sections from open PRs

APickledWalrus committed:
Add SectionUtils for common Trigger loading behavior

Absolutionism reviewed (approved):
Looks good to me. 
also I think EffZombify may be applicable for hint changing

sovdeeth reviewed (approved):
looking good to me

Efnilite commented on src/main/java/ch/njol/skript/lang/parser/ParserInstance.java:
```diff
@@ -533,6 +539,15 @@ public ExperimentSet getExperimentSet() {
+	private HintManager hintManager = new HintManager(true);
+
+	@ApiStatus.Experimental
+	public HintManager getHintManager() {
```
could use docs

Efnilite commented on src/main/java/ch/njol/skript/variables/HintManager.java:
```diff
@@ -0,0 +1,468 @@
+	/**
+	 * Copies hints from one scope to another.
+	 * Scopes are represented as integers, where {@code 0} represents the most recently entered scope
+	 * (i.e. the scope pushed by the most recent {@link #enterScope(boolean)} call).
```
this should also explain what int is used for the second most recent scope to establish what the pattern is, as it's not really clear on first read

Efnilite commented on src/main/java/ch/njol/skript/effects/EffSuppressTypeHints.java:
```diff
@@ -0,0 +1,61 @@
+@Name("Suppress Type Hints (Experimental)")
+@Description({
+	"An effect to suppress local variable type hint errors for the syntax lines that follow this effect.",
+	"NOTE: Suppressing type hints also prevents them from being collected."
```
collected where?

Efnilite commented on src/main/java/ch/njol/skript/variables/HintManager.java:
```diff
@@ -0,0 +1,468 @@
+ 	 * @param to The scope to copy hints to.
+	 * @param sectionOnly Whether only scopes representing sections should be considered.
+	 */
+	public void mergeScope(int from, int to, boolean sectionOnly) {
```
i feel like checking whether the ints are in bounds is a good idea here. same goes for the rest of the methods

Efnilite commented on src/main/java/ch/njol/skript/expressions/base/SectionExpression.java:
```diff
@@ -116,15 +125,18 @@ public final Section getAsSection() {
 	 * @return A trigger containing the loaded section. This should be stored and used
 	 * to run the section one or more times.
 	 */
 	protected Trigger loadCode(SectionNode sectionNode, String name,
```
should this have @SafeVarargs?

Efnilite commented on src/main/java/ch/njol/skript/lang/ExpressionSection.java:
```diff
@@ -83,8 +83,10 @@ public void setTriggerItems(List<TriggerItem> items) {
 
-	public Trigger loadCodeTask(SectionNode sectionNode, String name, @Nullable Runnable afterLoading, Class<? extends Event>... events) {
-		return super.loadCode(sectionNode, name, afterLoading, events);
+	public Trigger loadCodeTask(SectionNode sectionNode, String name,
```
same varargs here

Efnilite commented on src/main/java/ch/njol/skript/sections/EffSecShoot.java:
```diff
@@ -190,14 +190,11 @@ public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelaye
-				Skript.error("Delays cannot be used within a 'shoot' effect section");
-				return false;
-			}
+			trigger = SectionUtils.loadLinkedCode("shoot", (beforeLoading, afterLoading)
```
thoughts about adding a helper method that just has the node, name and event as arguments? that way it ensures nobody will call the wrong method without the runnables

APickledWalrus replied:
We aren't able to call the methods that build the triggers (e.g. `loadCode`) as they have restricted access (and it should probably remain that way). Unless you mean something else

Efnilite commented on src/main/java/ch/njol/skript/ScriptLoader.java:
```diff
@@ -983,11 +992,18 @@ public static ArrayList<TriggerItem> loadItems(SectionNode node) {
 				find_section:
 				try {
+					// enter capturing scope
+					// it is possible that the line may successful parse and initialize (via init), but some other issue
```
```suggestion
					// it is possible that the line may successfully parse and initialize (via init), but some other issue
```

APickledWalrus committed:
Improve documentation, safety checks

APickledWalrus committed:
Merge branch 'dev/feature' into feature/local-variable-type-hints
